spread/openstack.go

```diff
@@ -0,0 +1,758 @@
+	if !strings.Contains(addr, ":") {
+		addr += ":22"
+	}
+
+	// Iterate until the ssh connection to the host can be stablished
+	// In openstack the client cannot access to the serial console of the instance
+	timeout := time.After(openstackServerBootTimeout)
+	retry := time.NewTicker(openstackServerBootRetry)
+	defer retry.Stop()
+
+	for {
+		select {
+		case <-timeout:
+			return fmt.Errorf("cannot ssh to the allocated instance: timeout reached")
+		case <-retry.C:
+			_, err := sshDial("tcp", addr, config)
+			if err == nil {
+				debugf("Connection to server %s established", s.d.Name)
+				return nil
+			}
+		case <-ctx.Done():
+			return fmt.Errorf("cannot wait for %s to boot: interrupted", s)
+		}
+	}
+}
+
+func (p *openstackProvider) getSerialConsoleOutput(s *openstackServer) (string, error) {
+	url := fmt.Sprintf("servers/%s/action", s.d.Id)
+
+	var req struct {
+		OsGetSerialConsole struct{} `json:"os-getConsoleOutput"`
+	}
+	var resp struct {
+		Output string `json:"output"`
+	}
+	requestData := goosehttp.RequestData{ReqValue: req, RespValue: &resp, ExpectedStatus: []int{http.StatusOK}}
+
+	err := p.osClient.SendRequest("POST", "compute", "v2", url, &requestData)
+	if err != nil {
+		return "", err
+	}
+
+	return resp.Output, nil
+}
+
+func (p *openstackProvider) waitServerBootSerial(ctx context.Context, s *openstackServer) error {
+	timeout := time.After(openstackServerBootTimeout)
+	relog := time.NewTicker(60 * time.Second)
+	defer relog.Stop()
+	retry := time.NewTicker(openstackServerBootRetry)
+	defer retry.Stop()
+
+	var marker = openstackReadyMarker
+	for {
+		resp, err := p.getSerialConsoleOutput(s)
+		if err != nil {
+			return fmt.Errorf("cannot get console output for %s: %v", s, err)
+		}
+
+		if strings.Contains(resp, marker) {
+			return nil
+		}
+
+		select {
+		case <-retry.C:
+			debugf("Server %s is taking a while to boot...", s)
+		case <-relog.C:
+			printf("Server %s is taking a while to boot...", s)
+		case <-timeout:
+			return &FatalError{fmt.Errorf("cannot find ready marker in console output for %s: timeout reached", s)}
+		case <-ctx.Done():
+			return &FatalError{fmt.Errorf("cannot wait for %s to boot: interrupted", s)}
+		}
+	}
+	panic("unreachable")
+}
+
+func (p *openstackProvider) waitServerBoot(ctx context.Context, s *openstackServer) error {
+	printf("Waiting for %s to boot at %s...", s, s.address)
+	err := p.waitServerBootSerial(ctx, s)
+	if err != nil {
+		var fatalErr *FatalError
+		if errors.As(err, &fatalErr) {
+			return err
+		}
+		// It is important to try ssh connection because serial console could
+		// be disabled in the nova configuration
+		err = p.waitServerBootSSH(ctx, s)
+		if err != nil {
+			return fmt.Errorf("cannot connect to server %s: %v", s, err)
+		}
+	}
+	return nil
+}
+
+func (p *openstackProvider) address(ctx context.Context, s *openstackServer) (string, error) {
+	timeout := time.After(30 * time.Second)
+	retry := time.NewTicker(5 * time.Second)
+	defer retry.Stop()
+
+	for {
```

Comment:
@sergiocazzolato do we need this loop? does the IP take some time to be ready after the server is provisioned?

Reply:
yes, the address takes time to be associated to the instace, usually the instance is allocated and after few iterations it has an ip associated to its network interface